language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -47,9 +48,9 @@ infixr 9 `TypeSchemeArrow`
 -- @a@ is the Haskell denotation of a PLC type represented as a 'TypeScheme'.
 -- @r@ is the resulting type in @a@, e.g. the resulting type in
 -- @ByteString -> Size -> Integer@ is @Integer@.
-data TypeScheme a r where
-    TypeSchemeResult :: KnownType a => Proxy a -> TypeScheme a a
-    TypeSchemeArrow  :: KnownType a => Proxy a -> TypeScheme b r -> TypeScheme (a -> b) r
+data TypeScheme a costFun r where
+    TypeSchemeResult :: KnownType a => Proxy a -> TypeScheme a ExBudget a
+    TypeSchemeArrow  :: KnownType a => Proxy a -> TypeScheme b exB r -> TypeScheme (a -> b) (a -> exB) r
```

Comment:
I'll try to undo these changes.

Reply:
I tried 

```haskell
class CostingFun a r ex
instance {-# OVERLAPPABLE #-} ex ~ ExBudget => CostingFun r r ex
instance {-# OVERLAPPING #-} (CostingFun b r ex, aex ~ (a -> ex)) => CostingFun (a -> b) r aex
```

but that doesn't work, because Haskell lacks bidirectional instances. None of the type families approaches worked either.

We have two options:

1. merge as is, I'll fix the conflicts later. There are plenty of them, but most of them can be fixed by simply ignoring the changes to master on a per-file basis
2. use a type family and `unsafeCoerce` in a single place (where it's safe for the kind of builtins that we have):

```haskell
type family CostingFun a r where
    CostingFun (a -> b) r = a -> CostingFun b r
    CostingFun r        r = ExBudget

applyTypeSchemed
    :: forall e m a r. (MonadError (ErrorWithCause e) m, AsUnliftingError e, AsConstAppError e, SpendBudget m)
    => StagedBuiltinName -> TypeScheme a r -> a -> CostingFun a r -> [Value TyName Name ExMemory] -> EvaluateConstApp m ExMemory

<...>

                lift $ spendBudget (BBuiltin name) arg (unsafeCoerce exB)
```